File: packages/hono/tests/email-issuer.test.ts
import { describe, it, expect } from "vitest";
import { Hono } from "hono";
import Database from "better-sqlite3";
import { createAuth, runMigrations, defaultDeps, AuthError, type EmailSignIn } from "@mattsmith/passkey-sdk-core";
import { mountAuthRoutes } from "../src/index.js";

describe("host email issuer on mounted route", () => {
  for (const mode of ["success", "failure", "rejection", "empty", "invalid"] as const) {
    it(mode, async () => {
      const db = new Database(":memory:");
      try {
        runMigrations(db);
        let calls = 0;
        const signIn: EmailSignIn = async (input) => {
          calls++;
          expect(input).toMatchObject({ otpId: "pending", code: "123456", lifetimeSeconds: 3600, maxAttempts: 5, userAgent: "test-agent", ip: "192.0.2.1" });
          expect(input.now()).toBe(1700000000);
          if (mode === "failure") throw new Error("synthetic commit failure");
          if (mode === "rejection") throw new AuthError("otp_expired", "expired");
          if (mode === "empty") return { sessionToken: "", user: { id: "", email: "" } };
          return { sessionToken: "committed-token", user: { id: "user", email: "user@example.com" } };
        };
        const auth = createAuth({
          rpId: "example.com", origins: ["https://example.com"],
          session: { lifetimeSeconds: 3600, cookieName: "session" },
          email: { sendOtp: async () => {}, signIn },
          users: { findOrCreateByEmail: async () => { throw new Error("legacy resolver called"); } },
        }, { db, deps: { ...defaultDeps, now: () => 1700000000 } });
        // No OTP exists. Any accidental legacy verification would return 401.
        const app = new Hono(); mountAuthRoutes(app, auth);
        const res = await app.request("/auth/email/verify", {
          method: "POST", headers: { "content-type": "application/json", "user-agent": "test-agent", "x-forwarded-for": "192.0.2.1" },
          body: JSON.stringify({ otpId: "pending", code: mode === "invalid" ? "abc" : "123456" }),
        });
        expect(res.status).toBe(mode === "success" ? 200 : mode === "invalid" ? 400 : mode === "rejection" ? 410 : 500);
        expect(calls).toBe(mode === "invalid" ? 0 : 1);
        if (mode === "success") {
          expect(await res.json()).toEqual({ sessionToken: "committed-token", user: { id: "user", email: "user@example.com" } });
          expect(res.headers.get("set-cookie")).toContain("session=committed-token");
          expect(res.headers.get("set-cookie")).toContain("csrf=");
        } else expect(res.headers.get("set-cookie")).toBeNull();
        expect(db.prepare("SELECT COUNT(*) AS n FROM auth_sessions").get()).toEqual({ n: 0 });
      } finally { db.close(); }
    });
  }
});

File: servers/go/httpapi/email_issuer_test.go
package httpapi

import (
	"context"
	"errors"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/mattydsmith/passkey/servers/go/auth"
	"github.com/mattydsmith/passkey/servers/go/storage"
)

// Embedding a nil Storage makes any unintended legacy storage call panic.
// The mounted route must delegate before consuming an OTP or creating a session.
type unusedStorage struct{ storage.Storage }

func TestEmailIssuerMountedRoute(t *testing.T) {
	for _, mode := range []string{"success", "failure", "rejection", "empty", "invalid"} {
		t.Run(mode, func(t *testing.T) {
			calls := 0
			at := time.Unix(1700000000, 0)
			key := struct{}{}
			cfg := Config{Storage: unusedStorage{}, RPID: "example.com", Origins: []string{"https://example.com"}, SessionCookieName: "session", Now: func() time.Time { return at },
				GetOrCreateUserID: func(string) (string, error) { t.Fatal("legacy resolver called"); return "", nil },
				EmailSignIn: func(ctx context.Context, in EmailSignInInput) (EmailSignInResult, error) {
					calls++
					if ctx.Value(key) != "request" || in.OTPID != "pending" || in.Code != "123456" || in.MaxAttempts != auth.OTPMaxAttempts || in.SessionTTL != 30*24*time.Hour || !in.Now().Equal(at) || in.UserAgent == nil || *in.UserAgent != "test-agent" || in.IP == nil || *in.IP != "192.0.2.1" {
						t.Fatalf("input: %+v", in)
					}
					if mode == "failure" {
						return EmailSignInResult{}, errors.New("synthetic commit failure")
					}
					if mode == "rejection" {
						return EmailSignInResult{}, auth.ErrOTPExpired
					}
					if mode == "empty" {
						return EmailSignInResult{}, nil
					}
					return EmailSignInResult{SessionToken: "committed-token", User: EmailSignInUser{ID: "user", Email: "user@example.com"}}, nil
				},
			}
			r := chi.NewRouter()
			if err := Mount(r, cfg); err != nil {
				t.Fatal(err)
			}
			code := "123456"
			if mode == "invalid" {
				code = "abc"
			}
			req := httptest.NewRequest("POST", "/auth/email/verify", strings.NewReader(`{"otpId":"pending","code":"`+code+`"}`))
			req = req.WithContext(context.WithValue(req.Context(), key, "request"))
			req.Header.Set("User-Agent", "test-agent")
			req.Header.Set("X-Forwarded-For", "192.0.2.1")
			rec := httptest.NewRecorder()
			r.ServeHTTP(rec, req)
			status := 500
			wantCalls := 1
			switch mode {
			case "success":
				status = 200
			case "invalid":
				status = 400
				wantCalls = 0
			case "rejection":
				status = 410
			}
			if rec.Code != status || calls != wantCalls {
				t.Fatalf("status=%d calls=%d body=%s", rec.Code, calls, rec.Body.String())
			}
			if mode == "success" {
				if !strings.Contains(rec.Body.String(), `"sessionToken":"committed-token"`) || len(rec.Result().Cookies()) != 2 {
					t.Fatalf("success: %s cookies=%v", rec.Body.String(), rec.Result().Cookies())
				}
			} else if len(rec.Result().Cookies()) != 0 {
				t.Fatal("failure issued credentials")
			}
		})
	}
}
